docs/performance/remote-filesystem/amazon-s3.md

```diff
@@ -0,0 +1,396 @@
+you need to next override the desired package in the package group with a
+`Symfony\Component\Asset\UrlPackage` object that defines your base Amazon S3
+bucket URL.
+
+This is achived by extending `$app['asset.packages']`, and adding your specific
+`UrlPackage` object to he desired package pool name.
+
+
+#### Example
+
+This is an example of the `src/Provider/AssetServiceProvider.php` file.
+
+```php
+<?php
+
+namespace MySiteApp\Provider;
+
+use Silex\Application;
+use Silex\ServiceProviderInterface;
+use Symfony\Component\Asset\UrlPackage;
+
+class AssetServiceProvider implements ServiceProviderInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function register(Application $app)
+    {
+        $app['asset.packages'] = $app->share(
+            $app->extend(
+                'asset.packages',
+                function ($packages, $app) {
+                    $package = new UrlPackage(
+                        'http://example.s3-website.eu-central-1.amazonaws.com',
+                        $app['asset.version_strategy']('files'),
+                        $app['asset.context']
+                    );
+                    $packages->addPackage('files', $package);
+
+                    return $packages;
+                }
+            )
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function boot(Application $app)
+    {
+    }
+}
+```
+
+
+The first parameter for `UrlPackage` is the base URL or your Amazon S3 bucket.
+
+The second parameter is the version strategy applicable to the package, the
+value in the final parentheses **must** be the name of the package you're
+replacing.
+
+In this example, 'files' is package name, that is package that is responsible
+for the contents of `{web root}/files/`.
+
+Finally, you need to call `$packages->addPackage()` with the name of the
+package you're overriding, and the `UrlPackage` package object.
+
+
+### Filesystem
+
+#### Filesystem Cache Factory
+
+Create a protected callable `$app['filesystem.cache_factory']` that when
+invoked returns a class extending `\League\Flysystem\Cached\CachedAdapter`.
+
+Bolt provides the `\Bolt\Filesystem\Adapter\Cached` class that implements
+`CachedAdapter` for base needs.
+
+
+#### S3 Filesystem Service
+
+Create a protected callable `$app['filesystem.s3']` that when invoked will
+return an a class implementing `Bolt\Filesystem\FilesystemInterface`.
+
+If using `\Bolt\Filesystem\Filesystem` as your `FilesystemInterface` object,
+you first need to construct an `\Bolt\Filesystem\Adapter\S3` adapter class.
+
+Your adapter can then be passed to the `$app['filesystem.cache_factory']`
+callable, and the return value used as the adpater parameter to construct
+`Filesystem`.
+
+
+#### Extend Filesystem
+
+Finally, you must extend `$app['filesystem']`, and call `mountFilesystems()`
+using an assicative array of the filesystem(s) you want to override using the
+`$app['filesystem.s3']` protected callable.
```

Comment:
I would flip this order, so the user can follow why. "We want to override the "files" filesystem." -> We need an S3 adapter for that -> The adapter should be cached for performance.

Also you are mixing general documentation with specifics from the example below. Maybe it would be better to have those separated. Or maybe have more comments in the example below?